lnwallet/chanfunding/coin_select_test.go

```diff
@@ -530,3 +532,201 @@ func TestCoinSelectSubtractFees(t *testing.T) {
+		name: "spend far below",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value:    1 * coin,
+			},
+		}},
+		maxValue: 0.5 * coin,
+
+		// The one and only input will be selected.
+		expectedInput:      []btcutil.Amount{1 * coin},
+		expectedFundingAmt: 0.5 * coin,
+		expectedChange:     0.5*coin - fundingFee(feeRate, 1, true),
+	}, {
+		// We have 1.0 BTC available and want to spend just 1 Satoshi
+		// below that amount.
+		// This should lead to a funding TX with one output where the
+		// fee is subtracted from the total 1 BTC input value.
+		name: "spend little below",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value:    1 * coin,
+			},
+		}},
+		maxValue: 1*coin - 1,
+
+		// The one and only input will be selected.
+		expectedInput: []btcutil.Amount{
+			1 * coin,
+		},
+		expectedFundingAmt: 1*coin - fundingFee(feeRate, 1, false),
+		expectedChange:     0,
+	}, {
+		// The total funds available is below the dust limit after
+		// paying fees.
+		name: "dust output",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value: int64(
+					fundingFee(feeRate, 1, false) + dust,
+				),
+			},
+		}},
+		maxValue: fundingFee(feeRate, 1, false) + dust,
+
+		expectErr: true,
+	}, {
+		// If more than 20% of funds goes to fees, it should fail.
+		name: "high fee",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value: int64(
+					5 * fundingFee(feeRate, 1, false),
+				),
+			},
+		}},
+		maxValue: 5 * fundingFee(feeRate, 1, false),
+
+		expectErr: true,
+	}, {
+		// This test makes sure that the implementation detail of using
+		// CoinSelect and CoinSelectSubtractFees is done correctly.
+		// CoinSelect always defaults to use a fee for single input -
+		// one change tx, whereas CoinSelectSubtractFees will use a fee
+		// of single input - no change output, which without a sanity
+		// check could result in a local amount higher than the maximum
+		// amount that was expected.
+		name: "sanity check for correct maximum amount",
+		coins: []Coin{{
+			TxOut: wire.TxOut{
+				PkScript: p2wkhScript,
+				Value:    1 * coin,
+			},
+		}},
+		maxValue: 1*coin - fundingFee(feeRate, 1, false) - 1,
+
+		expectedInput:      []btcutil.Amount{1 * coin},
+		expectedFundingAmt: 1*coin - fundingFee(feeRate, 1, false) - 1,
+		expectedChange:     0,
+	}}
+
+	for _, test := range testCases {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			selected, localFundingAmt, changeAmt, err := CoinSelectUpToAmount(
+				feeRate, test.maxValue, dustLimit, test.coins,
+			)
+			if !test.expectErr && err != nil {
+				t.Fatalf(err.Error())
+			}
+
+			if test.expectErr && err == nil {
+				t.Fatalf("expected error")
+			}
+
+			// If we got an expected error, there is nothing
+			// more to test.
+			if test.expectErr {
+				return
+			}
+
+			// Check that the selected inputs match what we expect.
+			if len(selected) != len(test.expectedInput) {
```

Comment:
I know this code was written before we started using the `require` library, but it would be nice if we could use it for new PRs. Makes things much easier to read IMO.